### PR TITLE
Initial support for Python 3.12

### PR DESCRIPTION
**Pull request by DougBurke**
# Summary

Initial support for building with Python 3.12.

# Details

So

- `python setup.py arg` will now fail when `arg` is one of `test`, `develop`, `install`, or `build_sphinx` - reporting what the user should type instead
- the build has been restructured to handle the removal of distutils (and numpy.distutils) and it's not entirely clear how certain features should work (partly coupled with the setup changes I am about to mention), in particular the handling of the `stk.so` and `region.so` files which is **ugly** (in particular handling of editable mode vs normal installation)
  - see #50 
- configuration has been moved from `setup.cfg` to `pyproject.toml` as this is the "new" location (well, until PyPA changes things again)
  - we used to use `aliases` to include the sherpa/xspec config options into all the useful build options; we now hack these commands into the `build` structure and hope that it is sufficient (the setuptools documentation doesn't really describe all the options and how they now work)   
  - as part of this the minimum required setuptools version is now 64.0.0 (previously it was pinned < 60) because this is the first version with proper PEP 660 / editable build hook support (as we use pip we could relax it to >= 61.0.0) but given the large number of changes to setuptools and the number of bug fixes and changes in these releases we want a relatively-recent minimum version, just to avoid hitting build issues. Version 64.0.0 was released August 11 2022, so it should be readily available on all our supported platforms (for reference the current setuptools version is 69.0.3 released Dec 23 2023).
  - the `setup.cfg` file is now back to only being used to configure how to build parts of Sherpa

**NOTE**

I am not particularly happy with `MANIFEST.in` and the packages/data directory settings in `pyproject.tool`, but I have never been happy with them and I've decided that I really don't care any more. If anyone knows who to do this "properlly" then please by my guest! I may look at switching to a `src/` directory as this may make it easier, but I think Sherpa's current setup doesn't really map to the "simple" use cases supported by setuptools.

See also #1948, which was me playing around with a meson build. This worked, but I'd have needed to work out how to pass options into the meson build and it would have almost-certainly required changing the setup.cfg file format, which seems excessive.

# Examples

## building a wheel

With this I can use python 3.12 and say

```
% pip install build
% python -m build
... wait a long time
... then wait moar time as it redoes a lot
adding 'sherpa/utils/tests/test_utils_testing.py'
adding 'sherpa/utils/tests/test_utils_unit.py'
adding 'sherpa-4.16.0+26.gaf7f12fa.data/data/lib/python3.12/site-packages/group.so'
adding 'sherpa-4.16.0+26.gaf7f12fa.data/data/lib/python3.12/site-packages/stk.so'
adding 'sherpa-4.16.0+26.gaf7f12fa.dist-info/LICENSE'
adding 'sherpa-4.16.0+26.gaf7f12fa.dist-info/METADATA'
adding 'sherpa-4.16.0+26.gaf7f12fa.dist-info/WHEEL'
adding 'sherpa-4.16.0+26.gaf7f12fa.dist-info/entry_points.txt'
adding 'sherpa-4.16.0+26.gaf7f12fa.dist-info/top_level.txt'
adding 'sherpa-4.16.0+26.gaf7f12fa.dist-info/RECORD'
removing build/bdist.linux-x86_64/wheel
Successfully built sherpa-4.16.0+26.gaf7f12fa.tar.gz and sherpa-4.16.0+26.gaf7f12fa-cp312-cp312-linux_x86_64.whl
% pip list | grep sherpa
% pip install ./dist/sherpa-4.16.0+26.gaf7f12fa-cp312-cp312-linux_x86_64.whl
...
% pip list | grep sherpa
sherpa          4.16.0+26.gaf7f12fa
% cd ..
% sherpa_smoke
...
% sherpa_test
...
```

## building with XSPEC support

I've just tried a run with python 3.12, XSPEC 12.13.1, astropy, matplotlib, and DS9 and

```
% pip install -e .
% pytest
```

passed.

### CODE REVIEW

**Comment by DougBurke** [comment]
Rebased to pick up #1938

File: helpers/sherpa_config.py
Comment (DougBurke) [inline helpers/sherpa_config.py:178]: I need to drop this comment.

File: pyproject.toml
Comment (wmclaugh) [inline pyproject.toml:18]:
```suggestion
  { name = "Smithsonian Astrophysical Observatory / Chandra X-Ray Center", email = "cxchelp@cfa.harvard.edu" }
```

File: pyproject.toml
Comment (wmclaugh) [inline pyproject.toml:122]:
```suggestion
# It's not clear how this interacts with tools.setuptools.packages.
```

**Comment by DougBurke** [comment]
Tests currently fail because this PR doesn't contain #1957 (and probably #1954)

**Comment by DougBurke** [comment]
Rebased as #1954 and #1957 have now both been approved (and they should let the tests pass). Also addresses a comment I made.

**Comment by DougBurke** [comment]
Rebased to pick up recent build fixes.

**Comment by DougBurke** [comment]
Rebased to pick up recent build changes which lead to a conflict or two.

**Comment by anetasie** [comment]
@DougBurke should we include this PR in the release? I just tried to build this branch in python 3.12 on MacOS arm64 and It errored out with a lot of output lines which I marked as ..... in the screen output below.

```
% git status
HEAD detached at doug/test-setuptools-update
nothing to commit, working tree clean

% pip install .   
Processing /Users/aneta/git/sherpa
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
Requirement already satisfied: numpy in /Users/aneta/miniconda3-arm/envs/sherpa-py312/lib/python3.12/site-packages (from sherpa==4.16.0+126.ge4cc2334) (1.26.4)
Building wheels for collected packages: sherpa
  Building wheel for sherpa (pyproject.toml) ... error
  error: subprocess-exited-with-error
  
  × Building wheel for sherpa (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [2221 lines of output]
.......

      creating build/bdist.macosx-11.1-arm64/wheel/sherpa-4.16.0+126.ge4cc2334.data/data/lib/python3.12/site-packages
      error: can't copy '/Users/aneta/git/sherpa/build/lib/python3.12/site-packages/group.so': doesn't exist or not a regular file
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for sherpa
Failed to build sherpa
ERROR: Could not build wheels for sherpa, which is required to install pyproject.toml-based projects
```

**Comment by DougBurke** [comment]
@anetasie - **I** would like to get this in, but it requires **testing** and review by other people, and I don't know other people's schedules (I know CM do not have much spare time).

I do not have a macOS machine so I can provide no help with your various build issues (whether here or in other cases). I just don't know if the problem you noted above is with this PR or some more generic macOS build issue that you are having.

**Comment by anetasie** [comment]
@DougBurke the problem on my Mac is the same when building the main or this PR and I was not able to successfully build Sherpa in Python 3.12. I agree this needs more testing.

**Comment by DougBurke** [comment]
@hamogu are you able to build this PR on macOS?

**Review by hamogu — COMMENTED** [review]
In a new conda env with just Python 3.12 installed and nothing else (I expect that pip will install numpy and setuptools, since they are listed as  build requirements):
```
... a lot of outpt ...
      copying build/lib.macosx-11.0-arm64-cpython-312/sherpa/astro/models/_modelfcts.cpython-312-darwin.so -> build/bdist.macosx-11.0-arm64/wheel/sherpa/astro/models
      copying build/lib.macosx-11.0-arm64-cpython-312/sherpa/stats/_statfcts.cpython-312-darwin.so -> build/bdist.macosx-11.0-arm64/wheel/sherpa/stats
      running install_data
      error: can't copy '/Users/guenther/code/sherpa/build/lib/python3.12/site-packages/group.so': doesn't exist or not a regular file
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for sherpa
Failed to build sherpa
ERROR: Could not build wheels for sherpa, which is required to install pyproject.toml-based projects
```

I don't see that that error is necessarily Mac specific.

**Review by hamogu — COMMENTED** [review]
Wait, I had cashed files laying around from previous builds. After a good `git clean -dfx` (Ha! Dont' do that at home unluess you are sure you know wht's going to be deleted! I regret is every time I do) it builds (at least without XSPEC) but something is missing from the build:
```
(sherpa-py3.12) (base) MoritzAirRoseGold ~/code/sherpa> pytest
ImportError while loading conftest '/Users/guenther/code/sherpa/sherpa/conftest.py'.
sherpa/conftest.py:34: in <module>
    from sherpa.utils.err import RuntimeErr
sherpa/utils/__init__.py:39: in <module>
    from sherpa.utils._utils import hist1d, hist2d  # type: ignore
E   ModuleNotFoundError: No module named 'sherpa.utils._utils'
```

I've seen that error before but I don't remember now what the solution is. Either way, the fact that I remember it, means is not new form this PR - either something is still messed up in my enviroment, or the PR triggers a problem we already had in te past again. Just wrecking my brain now to remember what I did last time I saw that...

**Review by hamogu — COMMENTED** [review]
That's at least one argument to move more of the C++ functionality into Python (or use numpy where possible). It's not that our code is not good, it's that we don't have to worry about building it if we cna use numpy instead to make a histogram.

**Review by hamogu — COMMENTED** [review]
However, the compiled files clearly exisit in the right location (right next to the Python code that is being imported):
```

In [2]: import sherpa

In [3]: sherpa.__file__
Out[3]: '/Users/guenther/code/sherpa/sherpa/__init__.py'

In [4]: import sherpa.utils
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
Cell In[4], line 1
----> 1 import sherpa.utils

File ~/code/sherpa/sherpa/utils/__init__.py:39
     35 import numpy.fft
     37 # Note: _utils.gsl_fcmp and _utils.ndtri are not exported from
     38 #       this module; is this intentional?
---> 39 from sherpa.utils._utils import hist1d, hist2d  # type: ignore
     40 from sherpa.utils import _utils  # type: ignore
     41 from sherpa.utils import _psf    # type: ignore

ModuleNotFoundError: No module named 'sherpa.utils._utils'

In [5]:                                                                                                                                                                                           
Do you really want to exit ([y]/n)? 
(sherpa-py3.12) (base) MoritzAirRoseGold ~/code/sherpa> ls sherpa/utils
__init__.py                       _utils.cpython-312-darwin.so      formatting.py                     numeric_types.py                  src
__pycache__                       akima.py                          integration.cpython-312-darwin.so parallel.py                       testing.py
_psf.cpython-312-darwin.so        err.py                            logging.py                        random.py                         tests
```

(I'm posting comments here as I go along in the hope that it will be useful now or later to follow that path of thought - I appologize for extra noise for people who get email notification for every comment.)

**Comment by DougBurke** [comment]
Was this a 'pip install .' or a 'pip install -e .' ?

As noted in some of the build PRs (but with no issue to track this) I can get into cases where you get an error like the `ModuleNotFound` case, but - for me - it depends on some magical combination of stuff which I don't understand, replicate, or understand.....

**Review by hamogu — COMMENTED** [review]
For some reaons, I seem to have multiple versions of Python even in the same conda env. If I start it with ipython I get some version of python 3.10 (which fails, becuase the binary is compiled for Python 3.12). If I start a straight-up python, I get Pytohn 3.12 as requested.
That carries through to running the tests:

- `pytest` picks up Python 3.10, which fails
- `python -m pytest` pick up the correct Python 3.12 and passes (skipping or xfailing many of the tests because I don't have XSPEC or bokeh in this environment, but anyway, everything that's expected to pass, passes.)

So, bottom line is that this PR works - just my environment is still messed up in some inscrutable way.

**Review by hamogu — APPROVED** [review]
I can't say that I understand all parts of the installation process - @harlantc or @Marie-Terrell need to look at this. However, as far as I'm concerned, the proof is in the pudding. If the installation and tests work, it obviously works and we do need to support 3.12 now or latter - better now than later give that it's been out for a few months by now.
So, I'm approving this from my perspective, but I'd really encourage CM to take a look before it's merged.

**Comment by anetasie** [comment]
I followed @hamogu advice and run `git clean -dfx` in my python 3.12 conda environment. I can successfully build the branch now without XSPEC. I run `pytest` and have some warning msgs, but tests pass (exclude XSPEC and ds9 which were skipped).

```
sherpa % pytest sherpa/tests/test_data.py
=============================================== test session starts ===============================================
platform darwin -- Python 3.12.3, pytest-7.4.0, pluggy-1.0.0
rootdir: /Users/aneta/git/sherpa
configfile: pytest.ini
collected 571 items              
```
and a  long printout. Here is the Warning summary
 
```
================================================ warnings summary =================================================
../../miniconda3-arm/envs/sherpa-py312/lib/python3.12/site-packages/dateutil/tz/tz.py:37
  /Users/aneta/miniconda3-arm/envs/sherpa-py312/lib/python3.12/site-packages/dateutil/tz/tz.py:37: DeprecationWarning: datetime.datetime.utcfromtimestamp() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.fromtimestamp(timestamp, datetime.UTC).
    EPOCH = datetime.datetime.utcfromtimestamp(0)

<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: RuntimeWarning: numpy.ndarray size changed, may indicate binary incompatibility. Expected 80 from C header, got 96 from PyObject

sherpa/utils/testing.py:187
sherpa/utils/testing.py:187
  /Users/aneta/git/sherpa/sherpa/utils/testing.py:187: PytestUnknownMarkWarning: Unknown pytest.mark.xdist_group - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    grp = pytest.mark.xdist_group("ds9-tests")

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
```

**Comment by anetasie** [comment]
btw. this was building with `pip install -e  .`

**Comment by Marie-Terrell** [comment]
Note: CM plans to review this post 4.16.1 as we need to go through the implications of the changes here on all of the build types.

**Comment by DougBurke** [comment]
Rebased to address conflicts and pick up XSPEC 12.14.0 support.

**Comment by DougBurke** [comment]
Rebased to address conflict in documentation and to change several of the CI runs from Python 3.9 to 3.12.

**Comment by cdeil** [comment]
Super, thank you @DougBurke  and all!
It'll be great to have Sherpa on modern Python 3.12. :-)

**Comment by DougBurke** [comment]
Rebased to pick up recent changes and to check how it works with pip 24.2 / issue #2095

File: pyproject.toml
Comment (Marie-Terrell) [inline pyproject.toml:201]: This excluded the extra src directories in builds for me
```
[tool.setuptools.exclude-package-data]
"*" = ["src"]
```

**Comment by Marie-Terrell** [comment]
So, with this PR (and the additional pyproject.toml change), I now see the following file added to the install:
<site-packages path>/sherpa/include/sherpa/MersenneTwister.h
Do we want/need that header file (i.e. was it missing previously) or should this also be excluded?

File: pyproject.toml
Comment (DougBurke) [reply pyproject.toml:201]: So, this is one (of the many) areas of Python packaging I want to know less than what-little I already know. I know in my follow up to this PR - #2096 - where I try to move to a `src/` layout, I changed how this works (but I haven't tried to validate whether those changes are meaningful).

There may be an interaction with `MANIFEST.in` too, or maybe I'm just imaging it.

This is a long way of saying that I really don't have much understanding of this part of the build!

**Comment by Marie-Terrell** [comment]
With the exclude-package-data pyproject.toml change noted above, I ran the following tests:

sdist:
1.) (Current) Python 3.11, setuptools <60, numpy<2.0, and main (test-py311-setup-59)
2.) (setuptools update) Python 3.11, setuptools <60, numpy<2.0, and PR (test-py311-setup-new)
3.) (Python 3.12) Python 3.12, setuptools, numpy<2.0, and PR (test-py312-setup-new)
4.) (numpy) Python 3.12, setuptools, numpy, and PR (test-py312-numpy-new)

pip install:
(+matplotlib and using sdist above)
1.) (Current) Python 3.11, setuptools <60, numpy<2.0, and main
2.) (setuptools update) Python 3.11, setuptools, numpy<2.0, and PR
3.) (Python 3.12) Python 3.12, setuptools, numpy<2.0, and PR
4.) (numpy) Python 3.12, setuptools, numpy, and PR

(+matplotlib, cfitsio, xspec with sdist above and git source)
1.) (Current) Python 3.11, setuptools <60, numpy<2.0, and main
2.) (setuptools update) Python 3.11, setuptools, numpy<2.0, and PR
3.) (Python 3.12) Python 3.12, setuptools, numpy<2.0, and PR
4.) (numpy) Python 3.12, setuptools, numpy, and PR

(+matplotlib, astropy using sdist above)
1.) (Current) Python 3.11, setuptools <60, numpy<2.0, and main
2.) (setuptools update) Python 3.11, setuptools, numpy<2.0, and PR
3.) (Python 3.12) Python 3.12, setuptools, numpy<2.0, and PR
4.) (numpy) Python 3.12, setuptools, numpy, and PR


CIAO's Sherpa install
1.) Python 3.11 (Linux, macOS, macOS-ARM with numpy 1.24.* and setuptools>60)


Editable pip install
(Repeat above with "pip install -e")


CIAO's Sherpa Conda pkg:
1.) (Current) Python 3.11, setuptools <60, numpy<2.0, and main
2.) (setuptools update) Python 3.11, setuptools, numpy, and PR

Results: Running and comparing the tests to the currently expected results everything looks good. The only lingering change is the addition of one header file (noted above) to the install.

I can add a change to the sherpa conda recipe and the conda-deployment.yml to bump setuptools (and bump 3.9-3.11 -> 3.10-3.12 ?). Otherwise, when this merges the conda-deployment workflow will currently fail.

**Comment by DougBurke** [comment]
@Marie-Terrell - can you add the `pyproject.toml` change you need to this branch? I am also happy with dropping the python 3.12 / macOS issue as that has downstream consequences (which we can address later).

I still need a moment to look at the extra header file.

**Comment by DougBurke** [comment]
@Marie-Terrell I think we should not include `MersenneTwister.h` at this time

- this matches what we did before
- if people really wanted to use the Mersenne Twister algorithm they are better off using their own version of this file (is my thought at this time)

**Comment by Marie-Terrell** [comment]
Thank you, Doug. I can make a change to remove this along with the other files (as soon as I can). I am a bit held up here because in testing standalone Sherpa further to confirm the things in extern were installed correctly I noted that there are some issues. I believe this goes along with some of your previous notes on some of the group and etc. issues. I wasn't seeing issues because these are installed into the build, but are do not get into bin, lib, and top-level site-packages.

As we are not really supposed to be installing things outside "sherpa" in site-packages during a pip install (aside from entrypoints) I assume we are now running into issues with newer setuptools here. This is a bit of a showstopper on the change. I am not sure on the "correct" quickest path forward is. Effectively vendoring the libs under "sherpa" when not found to build against might be a possibility. I'll give this some thought, but we will likely need to have more conversations here. Just giving you an update

**Comment by DougBurke** [comment]
So, I tried 'python -m build' on both the current main branch and this PR, then listed the contents of the tar file and the wheel - since this PR is behind main there's a bunch of expected differences, but I get

# .tar.tg

Comparing the main branch to the PR gives (after dropping the leading sherpa-<versioninfo>/):

``
% diff tar.main tar.pr
6a7
> MANIFEST.in
146a148,149
> docs/_static/ui/
> docs/_static/ui/session_basic_example_fit.png
218d220
< docs/overview/astro_io_types.rst
4017a4020
> helpers/build.py
4020d4022
< helpers/develop.py
4027d4028
< notebooks/AsymmetricErrors.ipynb
4113,4114d4113
< sherpa/astro/tests/test_astro_data_asca_unit.py
< sherpa/astro/tests/test_astro_data_chandra_unit.py
4118d4116
< sherpa/astro/tests/test_astro_data_xmm_unit.py
4265a4264
> sherpa/optmethods/src/minim.cc
4373a4373
> sherpa/utils/src/gsl/fcmp.c
4422a4423,4429
> sherpa.egg-info/
> sherpa.egg-info/PKG-INFO
> sherpa.egg-info/SOURCES.txt
> sherpa.egg-info/dependency_links.txt
> sherpa.egg-info/entry_points.txt
> sherpa.egg-info/requires.txt
> sherpa.egg-info/top_level.txt
```

# wheel

```
% diff whl.main whl.pr
4,5d3
< group.so
< stk.so
57a56
> sherpa/astro/models/src/_modelfcts.cc
70,71d68
< sherpa/astro/tests/test_astro_data_asca_unit.py
< sherpa/astro/tests/test_astro_data_chandra_unit.py
75d71
< sherpa/astro/tests/test_astro_data_xmm_unit.py
109a106,114
> sherpa/astro/utils/src/PyWrapper.hh
> sherpa/astro/utils/src/_pileup.cc
> sherpa/astro/utils/src/_region.cc
> sherpa/astro/utils/src/_utils.cc
> sherpa/astro/utils/src/_wcs.cc
> sherpa/astro/utils/src/fftn.c
> sherpa/astro/utils/src/fftn.inc
> sherpa/astro/utils/src/pileup.cc
> sherpa/astro/utils/src/pileup.hh
116a122
> sherpa/astro/xspec/src/_xspec.cc
122a129,135
> sherpa/estmethods/src/estutils.cc
> sherpa/estmethods/src/estutils.hh
> sherpa/estmethods/src/estwrappers.cc
> sherpa/estmethods/src/info_matrix.cc
> sherpa/estmethods/src/info_matrix.hh
> sherpa/estmethods/src/projection.cc
> sherpa/estmethods/src/projection.hh
130a144
> sherpa/include/sherpa/MersenneTwister.h
155a170
> sherpa/models/src/_modelfcts.cc
173a189,202
> sherpa/optmethods/src/DifEvo.cc
> sherpa/optmethods/src/DifEvo.hh
> sherpa/optmethods/src/NelderMead.cc
> sherpa/optmethods/src/NelderMead.hh
> sherpa/optmethods/src/Opt.hh
> sherpa/optmethods/src/PyWrapper.hh
> sherpa/optmethods/src/RanOpt.hh
> sherpa/optmethods/src/Simplex.cc
> sherpa/optmethods/src/Simplex.hh
> sherpa/optmethods/src/_saoopt.cc
> sherpa/optmethods/src/minim.cc
> sherpa/optmethods/src/minim.hh
> sherpa/optmethods/src/minpack/LevMar.cc
> sherpa/optmethods/src/minpack/LevMar.hh
207a237
> sherpa/stats/src/_statfcts.cc
245a276,325
> sherpa/utils/src/_psf.cc
> sherpa/utils/src/_utils.cc
> sherpa/utils/src/integration.cc
> sherpa/utils/src/cephes/cephes.h
> sherpa/utils/src/cephes/const.c
> sherpa/utils/src/cephes/fabs.c
> sherpa/utils/src/cephes/gamma.c
> sherpa/utils/src/cephes/igam.c
> sherpa/utils/src/cephes/igami.c
> sherpa/utils/src/cephes/incbet.c
> sherpa/utils/src/cephes/incbi.c
> sherpa/utils/src/cephes/isnan.c
> sherpa/utils/src/cephes/mconf.h
> sherpa/utils/src/cephes/mtherr.c
> sherpa/utils/src/cephes/ndtri.c
> sherpa/utils/src/cephes/polevl.c
> sherpa/utils/src/cephes/protos.h
> sherpa/utils/src/gsl/err.c
> sherpa/utils/src/gsl/error.c
> sherpa/utils/src/gsl/fcmp.c
> sherpa/utils/src/gsl/fcmp.h
> sherpa/utils/src/gsl/gsl_errno.h
> sherpa/utils/src/gsl/gsl_integration.h
> sherpa/utils/src/gsl/gsl_machine.h
> sherpa/utils/src/gsl/gsl_math.h
> sherpa/utils/src/gsl/gsl_message.h
> sherpa/utils/src/gsl/gsl_nan.h
> sherpa/utils/src/gsl/gsl_pow_int.h
> sherpa/utils/src/gsl/gsl_precision.h
> sherpa/utils/src/gsl/gsl_sys.h
> sherpa/utils/src/gsl/gsl_types.h
> sherpa/utils/src/gsl/message.c
> sherpa/utils/src/gsl/qng.c
> sherpa/utils/src/gsl/qng.h
> sherpa/utils/src/gsl/stream.c
> sherpa/utils/src/gsl/strerror.c
> sherpa/utils/src/sjohnson/Faddeeva.cc
> sherpa/utils/src/sjohnson/Faddeeva.hh
> sherpa/utils/src/sjohnson/adapt_integrate.c
> sherpa/utils/src/sjohnson/adapt_integrate.h
> sherpa/utils/src/tcd/tcd.h
> sherpa/utils/src/tcd/tcdCastArray.c
> sherpa/utils/src/tcd/tcdError.c
> sherpa/utils/src/tcd/tcdFFTConvolve.c
> sherpa/utils/src/tcd/tcdInitConvolveOut.c
> sherpa/utils/src/tcd/tcdInitTransform.c
> sherpa/utils/src/tcd/tcdPadData.c
> sherpa/utils/src/tcd/tcdPixelArith.c
> sherpa/utils/src/tcd/tcdTransform.c
> sherpa/utils/src/tcd/tcd_private.h
259,264c339,346
< sherpa-4.16.1+298.g536f938a.dist-info/LICENSE
< sherpa-4.16.1+298.g536f938a.dist-info/METADATA
< sherpa-4.16.1+298.g536f938a.dist-info/WHEEL
< sherpa-4.16.1+298.g536f938a.dist-info/entry_points.txt
< sherpa-4.16.1+298.g536f938a.dist-info/top_level.txt
< sherpa-4.16.1+298.g536f938a.dist-info/RECORD
---
> sherpa-4.16.1+173.gd0c32478.data/data/lib/python3.11/site-packages/group.so
> sherpa-4.16.1+173.gd0c32478.data/data/lib/python3.11/site-packages/stk.so
> sherpa-4.16.1+173.gd0c32478.dist-info/LICENSE
> sherpa-4.16.1+173.gd0c32478.dist-info/METADATA
> sherpa-4.16.1+173.gd0c32478.dist-info/WHEEL
> sherpa-4.16.1+173.gd0c32478.dist-info/entry_points.txt
> sherpa-4.16.1+173.gd0c32478.dist-info/top_level.txt
> sherpa-4.16.1+173.gd0c32478.dist-info/RECORD
```

For the wheel

a) there's a whole bunch of code needed to compile the code which I **don't think** but am not completely sure is unnescary
b) you can see the `group.so`/`stk.so` changes. Looking at the compiled files you see

## main branch

```
%  unzip -l old/sherpa-4.16.1+298.g536f938a-cp311-cp311-linux_x86_64.whl | grep \.so
   249080  09-18-2024 19:21   group.so
    85200  09-18-2024 19:21   stk.so
   175816  09-18-2024 19:22   sherpa/astro/models/_modelfcts.cpython-311-x86_64-linux-gnu.so
    44344  09-18-2024 19:22   sherpa/astro/utils/_pileup.cpython-311-x86_64-linux-gnu.so
   442232  09-18-2024 19:22   sherpa/astro/utils/_region.cpython-311-x86_64-linux-gnu.so
    66888  09-18-2024 19:22   sherpa/astro/utils/_utils.cpython-311-x86_64-linux-gnu.so
   908232  09-18-2024 19:22   sherpa/astro/utils/_wcs.cpython-311-x86_64-linux-gnu.so
    45920  09-18-2024 19:22   sherpa/estmethods/_est_funcs.cpython-311-x86_64-linux-gnu.so
   129976  09-18-2024 19:22   sherpa/models/_modelfcts.cpython-311-x86_64-linux-gnu.so
   187624  09-18-2024 19:22   sherpa/optmethods/_saoopt.cpython-311-x86_64-linux-gnu.so
   154248  09-18-2024 19:22   sherpa/optmethods/_tstoptfct.cpython-311-x86_64-linux-gnu.so
     4445  08-09-2024 19:28   sherpa/optmethods/tests/test_ncoresopt.py
    55104  09-18-2024 19:22   sherpa/stats/_statfcts.cpython-311-x86_64-linux-gnu.so
    74320  09-18-2024 19:22   sherpa/utils/_psf.cpython-311-x86_64-linux-gnu.so
   142432  09-18-2024 19:22   sherpa/utils/_utils.cpython-311-x86_64-linux-gnu.so
    40096  09-18-2024 19:22   sherpa/utils/integration.cpython-311-x86_64-linux-gnu.so
```

## this PR

```
% unzip -l new/sherpa-4.16.1+173.gd0c32478-cp311-cp311-linux_x86_64.whl | grep \.so
   175816  09-18-2024 19:24   sherpa/astro/models/_modelfcts.cpython-311-x86_64-linux-gnu.so
    44344  09-18-2024 19:24   sherpa/astro/utils/_pileup.cpython-311-x86_64-linux-gnu.so
   442232  09-18-2024 19:24   sherpa/astro/utils/_region.cpython-311-x86_64-linux-gnu.so
    75272  09-18-2024 19:24   sherpa/astro/utils/_utils.cpython-311-x86_64-linux-gnu.so
   908232  09-18-2024 19:24   sherpa/astro/utils/_wcs.cpython-311-x86_64-linux-gnu.so
    45920  09-18-2024 19:24   sherpa/estmethods/_est_funcs.cpython-311-x86_64-linux-gnu.so
   129976  09-18-2024 19:24   sherpa/models/_modelfcts.cpython-311-x86_64-linux-gnu.so
   187648  09-18-2024 19:24   sherpa/optmethods/_saoopt.cpython-311-x86_64-linux-gnu.so
   154248  09-18-2024 19:24   sherpa/optmethods/_tstoptfct.cpython-311-x86_64-linux-gnu.so
     4445  07-05-2024 14:20   sherpa/optmethods/tests/test_ncoresopt.py
    55104  09-18-2024 19:24   sherpa/stats/_statfcts.cpython-311-x86_64-linux-gnu.so
  1164480  09-18-2024 19:24   sherpa/utils/_psf.cpython-311-x86_64-linux-gnu.so
   138304  09-18-2024 19:24   sherpa/utils/_utils.cpython-311-x86_64-linux-gnu.so
    40096  09-18-2024 19:24   sherpa/utils/integration.cpython-311-x86_64-linux-gnu.so
   126227  07-05-2024 14:20   sherpa/utils/src/sjohnson/Faddeeva.cc
     2645  07-05-2024 14:20   sherpa/utils/src/sjohnson/Faddeeva.hh
    21940  07-05-2024 14:20   sherpa/utils/src/sjohnson/adapt_integrate.c
      559  07-05-2024 14:20   sherpa/utils/src/sjohnson/adapt_integrate.h
   249088  09-18-2024 19:24   sherpa-4.16.1+173.gd0c32478.data/data/lib/python3.11/site-packages/group.so
    85208  09-18-2024 19:24   sherpa-4.16.1+173.gd0c32478.data/data/lib/python3.11/site-packages/stk.so
```

**Comment by Marie-Terrell** [comment]
@hamogu and @DougBurke Please take a look at the message above.
Questions summarized (see comment above for details):
- Pip install new warning about *.src "packages": I assume we want to caveat this and fix post-release with namespace changes or the work Doug is doing with src restructuring?
- Conda package content changes (removal of extra files from export/ previously included):
  - Why are those files there previously? 
  - Do they need to be for "normal" sherpa use?
  - Are they needed for any edge cases?
  - Is an additional comment needed in the "install_dir" line in setup.cfg to note that a full-path is not supported/intended?
Note: The conda package content change is specific to standalone Sherpa and does not change CIAO's Sherpa.

**Comment by hamogu** [comment]
**pip install warning:** I don't think we have to worry about a warning that we partially (mostly?) understand and that only shows up with the `-vv` option. I didn't even know that that option existed. [And I also agree with the sentiment in the linked astropy issue that it may not be the best use of our time to make some tool happy that might change its defaults on the next version. We can change it when it becomes a problem.]

 **"Is an additional comment needed"**: Comments never hurt, but very few people will care - I doubt that anyone except the people on this PR will ever notice. The only exception to that would be if sherpa gets packaged by someone for inclusion
in a linux distro, e.g. build as a Debian package. In that case, that person would want to unbundle stk, gap and friends and do some things that we would consider a very non-standard setup. However, to my knowledge that's not happening - it if was we'd probably see issues from that team already. 

Everybody else either installs from conda or pip - and we test and know that those work.


**Conda package content**: I have no idea why those files were in the conda packages before. My gut feeling is that nobody ever thought about it carefully and including a few unused files didn't do any harm so we never noticed. Similar to the above, I'm not worried about edge cases. Very few people have those edge cases and those that do as usually competent in building with pip so we can ask them to do that (possibly after a fix to `main` if needed to support some valid edge case).

### Summary
I think that the benefits of being able to support 3.12 far outweigh any risks with this PR. I've already spend two hours trying to figure out why my sherpa pip build in some environment didn't work before I realized that I had just pulled the default Python form conda, which is now 3.12. That's going to happen more and more; it's certainly far more common than the edge cases you mentioned above. Importantly, downstream packages (like BXA and others) are also getting stuck with lower versions if we don't update.

**Comment by Marie-Terrell** [comment]
Also to note: We can remove the remnants of Python 3.9 tests (and unsupport it?) and add the Python 3.12 conda package builds in a follow-up PR.

**Comment by DougBurke** [comment]
For **pip install warning** 

- I think we can just "ignore it" for now; I'm not even sure it's worth documenting
- #2096 is my latest attempt to "address" this, but I ran out of steam and have never decided whether it's worth it or not 

Historically there has been a lot of back and forth over the "use `src/` versus how we do it", and I can see some benefit to having a more-obvious separation that `src/` would provide,  which is why I keep on coming back to it, but it's never been a high priority. But maybe we are close to being able to do it!

**Comment by DougBurke** [comment]
For **conda package differences**

- I also have no idea about what/why/how the conda build is doing what it's doing, and how that interacts with the python build system
- I say we go out with the new version (which, as @Marie-Terrell points out, seems to be working fine) and if it causes problems for downstream users then we can worry about it then

**Review by DougBurke — COMMENTED** [review]
I would love to approve this, but GitHub quite rudely claims that "Pull request authors can't approve their own pull requests"...

**Review by Marie-Terrell — APPROVED** [review]
Looks good!

**Comment by DougBurke** [comment]
Thanks everyone for this **long** journey.

**Pull request merged**
